### PR TITLE
Port obtaining server info from srb2kartjs

### LINKED ISSUE #4: SRB2KartJS port: Network from srb2kartserverinfo.js

### PR DESCRIPTION
**Pull request by Callmore**
Closes #4

### CODE REVIEW

**Comment by Fl0GUI** [comment]
Set me as a reviewer when you're ready or want feedback

File: src/srb2kart/serverinfo/serverinfo.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+type packetType uint8
+
+const (
+	packetTypeServerInfo packetType = 0x0D
+	packetTypePlayerInfo packetType = 0x0E
+)
+
+type KartPacket interface{
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: There is no need for this to be public

File: src/srb2kart/serverinfo/serverinfo.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+	GetPacketType() packetType
+}
+
+type KartPacketHeader struct {
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: No need for public, it's only applicable in this package

File: src/srb2kart/serverinfo/serverinfo.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+	MapName        string
+	MapTitle       string
+	MapMD5         [16]uint8
+	ActNum         uint8
+	IsZone         uint8
+	HttpSource     string
+	FileNeeded     []FileNeededEntry
+}
+
+func (p *KartServerInfoPacket) GetPacketType() packetType {
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: Can be private

File: src/srb2kart/serverinfo/serverinfo.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+type KartPlayerInfoPacket struct {
+	Header     KartPacketHeader
+	PlayerInfo []KartPlayerInfoEntry
+}
+
+func (p *KartPlayerInfoPacket) GetPacketType() packetType {
+	return p.Header.PacketType
+}
+
+func nullTerminated(data []byte) string {
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: Can be in the functional/strings package

File: src/srb2kart/serverinfo/serverinfo.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+	newBytes := make([]byte, 0)
+
+	// Filter out all bytes above 127
+	for i := 0; i < len(data); i++ {
+		if data[i] <= 127 {
+			newBytes = append(newBytes, data[i])
+		}
+	}
+	before, _, _ := bytes.Cut(newBytes, []byte{0})
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: You can know the index where the loop ended. No need to look for the separator

File: src/srb2kart/serverinfo/serverinfo.go [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+		entry.FileName, _ = buf.ReadString(0)
+		// Trim off the null terminator
+		entry.FileName = entry.FileName[:len(entry.FileName)-1]
+		binary.Read(buf, binary.LittleEndian, &entry.MD5)
+		entries = append(entries, entry)
+	}
+	return entries
+}
+
+func unpackServerInfoPacket(data []byte) (*KartServerInfoPacket, error) {
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: Imo if the only reason to return a pointer is to be able to return nil on error, while the function also constructs the type, it's better to return the value and on error return an uninitialised value.

File: src/srb2kart/serverinfo/serverinfo.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+			Skin:         entry.Skin,
+			Data:         entry.Data,
+			Score:        entry.Score,
+			TimeInServer: entry.TimeInServer,
+		})
+	}
+	return &packet, nil
+}
+
+func GetSRB2Info(adress string) (*KartServerInfoPacket, *KartPlayerInfoPacket, error) {
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: This should be the only public function. 
But also, why pointers. If I can avoid those I will.

File: src/srb2kart/serverinfo/serverinfo.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
+package serverinfo
```
Comment (Fl0GUI) [inline src/srb2kart/serverinfo/serverinfo.go]: Here's how I envisioned the package structure:
GoBun/srb2kart/network
  Contains all the basic functions, these pass the call towards the package specific parsers
GoBun/srb2kart/network/serverinfo
  Contains the content of this file

**Review by Fl0GUI — CHANGES_REQUESTED** [review]
Please rework how the code is structured. Eg. Group the raw packages with the request package, the parse functions, etc. 
If you can think of a name for two things, that can be the filename where those two things are.

**Comment by Fl0GUI** [comment]
Oh also make a script in src/scripts. It can be a very simple one, one art for the address, or default to localhost:5029, and print out the two packages

File: src/srb2kart/serverinfo/serverinfo.go (hunk 2 above)
Comment (Callmore) [reply src/srb2kart/serverinfo/serverinfo.go]: Should the `Header` field in the public packet structs remain public if this is made private?

File: src/srb2kart/serverinfo/serverinfo.go (hunk 8 above)
Comment (Callmore) [reply src/srb2kart/serverinfo/serverinfo.go]: So would `GoBun/srb2kart/network` effectively contain the public parts of the current `serverinfo` package and `GoBun/srb2kart/network` would contain the private parts?

File: src/srb2kart/serverinfo/serverinfo.go (hunk 2 above)
Comment (Fl0GUI) [reply src/srb2kart/serverinfo/serverinfo.go]: I wouldn't think so. The header is only important for the networking thing. The body is for the user of the package.

File: src/srb2kart/serverinfo/serverinfo.go (hunk 8 above)
Comment (Fl0GUI) [reply src/srb2kart/serverinfo/serverinfo.go]: I'd say that the body structs can be part of `GoBun/srb2kart/network/info` (there might be a better name) while the public function, `GetSRB2Info`, is public in `GoBun/srb2kart/network`

**Comment by Callmore** [comment]
memory is hard

File: src/functional/strings/strings.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package strings
+
+// NullTerminatedString returns a string from a null-terminated byte slice,
+// removing any bytes over 127.
```
Comment (Fl0GUI) [inline src/functional/strings/strings.go:4]: Minor comment, but what about unicode characters? If you can't find some east "IsPrintable" check feel free to ignore

File: src/scripts/serverinfo/getserverinfo.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"GoBun/srb2kart/network"
+	"fmt"
+	"os"
+	"strings"
+)
+
+func coloriseServerName(name []byte) string {
```
Comment (Fl0GUI) [inline src/scripts/serverinfo/getserverinfo.go:10]: If you're going to define this twice might as well put it in `GoBun/strings`

File: src/scripts/getserverlist/getserverlist.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+				output.WriteString("\x1b[38;5;217m")
+			case 142: // Brown
+				output.WriteString("\x1b[38;5;94m")
+			case 143: // Peach
+				output.WriteString("\x1b[38;5;215m")
+			}
+		} else {
+			// Otherwise, append the byte
+			output.WriteByte(b)
+		}
```
Comment (Fl0GUI) [inline src/scripts/getserverlist/getserverlist.go:63]:
```go
func colorCode(b byte) {
    var lookup = []string{...}
    return lookup[b - 128]
}

if b >= 128 {
    colorCode(b)
} else {
   b
}
```

but with the appropriate bound checks

File: src/scripts/serverinfo/getserverinfo.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+func main() {
+	if len(os.Args) < 2 {
+		println("Usage: getserverinfo <ip:port>")
+		os.Exit(1)
+	}
+
+	serverIp := os.Args[1]
+
+	// Add a port if it's not there
+	if !strings.Contains(serverIp, ":") {
```
Comment (Fl0GUI) [inline src/scripts/serverinfo/getserverinfo.go:75]: You could use the `net/http` standard package for this

File: src/srb2kart/network/info/info.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+type packetHeader struct {
+	Checksum   uint32
+	Ack        uint8
+	Ackret     uint8
+	PacketType PacketType
+	_          [1]byte
+}
+
+type serverInfoPacketRaw struct {
```
Comment (Fl0GUI) [inline src/srb2kart/network/info/info.go:37]: I'd still put the server info and player info stuff in separate files, and the common types in another.

Comment (Fl0GUI) [reply src/srb2kart/network/info/info.go:37]: like, if you were to create the path to the file, you'd end up with `.../network/info/info.go`, which is redundant. There's more useful information about the code that you can put in the filename to organize the code better.

**Review by Fl0GUI — CHANGES_REQUESTED** [review]
I'd still like to see plenty of refactors, but it's good enough to be merged. If you want to merge this and let me refactor it someday that's fine.

**Pull request merged**